[Monitoring] Metricbeat migration for logstash, beats and apm

chrisronline opened the pull request:
This PR is step 3 of the [Metricbeat setup wizard](https://github.com/elastic/kibana/issues/39010).

It adds MB migration UX for logstash, beats and APM, including net new user experience as well as migration from internal collection.

This PR builds upon the functionality added in the two step 2 PRs, [part 1](https://github.com/elastic/kibana/pull/35228) and [part 2](https://github.com/elastic/kibana/pull/39832).

In addition, this PR also adds a button to the bottom of each stack products' listing page that allows the user to access the setup wizard to add monitoring for a new node/instance:
<img width="406" alt="kibana" src="https://user-images.githubusercontent.com/56682/61640913-2349df00-ac6c-11e9-87a8-8e6c00e96953.png">
<img width="374" alt="logstash" src="https://user-images.githubusercontent.com/56682/61640914-2349df00-ac6c-11e9-995d-3184648e21ba.png">
<img width="384" alt="beats" src="https://user-images.githubusercontent.com/56682/61640915-2349df00-ac6c-11e9-82eb-5645ea7314e6.png">
<img width="390" alt="apm" src="https://user-images.githubusercontent.com/56682/61640916-2349df00-ac6c-11e9-9392-01d6a931bbed.png">
<img width="450" alt="es" src="https://user-images.githubusercontent.com/56682/61640917-23e27580-ac6c-11e9-858d-82727991825d.png">

Also, this PR takes advantage of the fact that we're attempting to detect if certain stack products exist, based on known indices. This wasn't something that mattered in previous PRs as we can know for certain that a Kibana and ES exist if they are in the stack monitoring UI.

<img width="1673" alt="logstash" src="https://user-images.githubusercontent.com/56682/61642383-cac81100-ac6e-11e9-8cc9-ff9a36f818e1.png">
<img width="1622" alt="beats" src="https://user-images.githubusercontent.com/56682/61642384-cac81100-ac6e-11e9-841f-369ac31c668d.png">
<img width="1628" alt="apm" src="https://user-images.githubusercontent.com/56682/61642385-cac81100-ac6e-11e9-8d83-0bdcbd8060f5.png">


Lastly, this PR adds code to disable entering setup mode if on cloud.

During testing, we realized that logstash and beats can potentially contain mb monitoring documents without a `cluster_uuid` (which would put those instances in the [standalone cluster](https://github.com/elastic/kibana/pull/28278)) so we built a flow in the wizard to let the user know the location of this instance/node will change. In this flow, the user is required to check the checkbox to finish the wizard. Here is what that looks like:
<img width="929" alt="Screen Shot 2019-07-26 at 11 30 48 AM" src="https://user-images.githubusercontent.com/56682/61963341-5a383180-af99-11e9-860b-fe4338145b39.png">
<img width="918" alt="Screen Shot 2019-07-26 at 11 30 42 AM" src="https://user-images.githubusercontent.com/56682/61963342-5a383180-af99-11e9-8674-d075e7c25134.png">


## Testing
The first thing to talk about is enabling this "Setup Mode". The most straight forward way is to just update the query string in the url, by adding this to the global state param: `,inSetupMode:!t`. A final, working URL will look something like: http://localhost:5601/app/monitoring#/elasticsearch/nodes?_g=(cluster_uuid:qtoNYcCxQdK-9BGIdPN_lw,inSetupMode:!t)

- Please test all three newly supported stack products: Logstash, Beats, and APM
- For each stack product, please test both flows: net new user (no existing monitoring data for the stack product) and migration (existing, internally collected monitoring data)
- For net new user flow, there are two potential sub-flows
  - We _detect_ the stack product _might_ exist and provide messaging to that effect
  - We are unable to detect any product and the only option is the add monitoring button at the bottom of the listing page
- For APM, follow these steps:
  1) Clone and run https://github.com/elastic/apm-server following instructions in README
  2) Install the apm client within Kibana itself
     1) `yarn add elastic-apm-node`
     2) Add the following code to the top of the file`scripts/kibana.js`: `require('elastic-apm-node').start({})`

## Existing Issues
- I ran into an issue with Beats in the migration flow, specifically in the disable internal collection step. Beats is designed to only make the connection to the ES output (assuming it has an ES output configured) once data is ready to be sent. It's possible that in the process of restarting the beat (after disabling internal collection) will not yield a new event, so the ES output connection is never registered which causes issues in this flow (as MB will still collect from the beats API but the beat API will not report a valid `cluster_uuid` as a result). This can be addressed in testing by deleting the `data` directory in the beat and restarting.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5944/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6970/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7468/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7530/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/7489/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8266/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8453/)

spalger commented:
retest

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8481/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8682/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8707/)

elasticmachine commented:
Pinging @elastic/stack-monitoring

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8728/)

ycombinator commented:
<img width="495" alt="Screen Shot 2019-07-22 at 4 57 55 PM" src="https://user-images.githubusercontent.com/51061/61672790-df71cc80-aca1-11e9-8127-3f904474f32b.png">

Should say Beats and beat instead of Nodes and node, respectively, in the last sentence?

Also, regarding this:

> if there are more detected ... then ...

Can't the code just do this detection for the user and present the appropriate message directly, rather than asking the user to do the comparison and make the appropriate choice?

ycombinator commented:
Related, in that I noticed this while reviewing this PR (but not directly a result of this PR): https://github.com/elastic/kibana/issues/41725.

ycombinator commented:
<img width="398" alt="Screen Shot 2019-07-22 at 5 27 50 PM" src="https://user-images.githubusercontent.com/51061/61673863-13e78780-aca6-11e9-86f6-5eb3c7af2323.png">

Ideally we wouldn't say "a few seconds" for the uptime if there was no monitoring data for any logstash nodes.

ycombinator commented:
When the user is in Setup Mode and we have no monitoring data for Logstash or Beats or APM Server, should we disable access to the following pages?

Logstash > Overview
Logstash > Pipelines
Beats > Overview
APM > Overview

ycombinator commented:
<img width="968" alt="Screen Shot 2019-07-22 at 5 35 56 PM" src="https://user-images.githubusercontent.com/51061/61674093-2d3d0380-aca7-11e9-9ce0-7f26935e45b6.png">
<img width="1123" alt="Screen Shot 2019-07-22 at 5 34 06 PM" src="https://user-images.githubusercontent.com/51061/61674095-2e6e3080-aca7-11e9-9fbe-11082d07ea81.png">

Inconsistent language between Logstash and Beats for the "might exist" scenario. I know you're cleaning up language later so if you want to defer this to that stage, that's fine with me.

chrisronline commented:
@ycombinator I've addressed your feedback, as well as resolving https://github.com/elastic/kibana/issues/41725 in this PR too. Ready for more 👀

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8969/)

ycombinator commented:
> as well as resolving #41725 in this PR too

@chrisronline Could you please put up a separate PR for these changes? In part because it will make this PR here easier to review and also in part because there's still active discussion going on in #41725. That way we don't have to block this PR on any discussions for #41725.

chrisronline commented:
@ycombinator Partially done. I moved the kibana and ES panel changes to [this PR](https://github.com/elastic/kibana/pull/41898), but I'm going to hold off on applying these changes to the beats/logstash/apm until the previously mentioned PR is merged. I've updated the meta to reflect this too

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9161/)

ycombinator commented:
Re: https://github.com/elastic/kibana/pull/40442#issuecomment-514005942, I'm still able to get to the disabled pages by clicking on the enabled pages and then clicking on the other tabs.

ycombinator commented:
I setup one Logstash node with internal monitoring. When I went to to the Node Listing page, I verified that the Setup Status was set to "Internal collection" and the Migrate button was shown.

I clicked on the Migrate button to start the migration. I followed the instructions and clicked the "Check for data" button. The UI started checking for data every 5 seconds. Even after a minute or so, I still see this message:

> We have not detected any monitoring data coming from Metricbeat for this Logstash. We will continuously check every 5 seconds in the background.

I opened Console in a new tab and ran the following query:

```
POST .monitoring-logstash-7-mb-*/_search
{
  "_source": [
    "timestamp",
    "type"
  ],
  "collapse": {
    "field": "type"
  },
  "sort": [
    {
      "timestamp": {
        "order": "desc"
      }
    }
  ]
}
```

I can confirm that I'm seeing `logstash_stats` documents within the last 20 seconds  and one `logstash_state` document from when I started up Metricbeat.

Any ideas why the Setup UI isn't detecting monitoring data sent by Metricbeat?

chrisronline commented:
@ycombinator Nice catch. I've updated all the links now to be disabled, but in the process, realized the previous code to disable them, didn't actually disable them, but just made them _look disabled_. I fixed that up too

chrisronline commented:
@ycombinator I'm not able to reproduce that. Are you able to reproduce it consistently?

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9220/)

ycombinator commented:
> I'm not able to reproduce that. Are you able to reproduce it consistently?

I am, but it's possible I wasn't in a clean state. Let me try again after building everything from scratch and deleting all data folders.

ycombinator commented:
I have just internal monitoring setup for 1 Logstash instance. Haven't yet started the process of migrating it to Metricbeat. I'm in Setup Mode and started on the Cluster Overview page. Then I clicked the Logstash Node Listing link to start migrating to Metricbeat. I noticed that when I go to the Logstash Node Listing page the Logstash Overview tab disappears.

![Jul-25-2019 08-34-10](https://user-images.githubusercontent.com/51061/61887980-88454500-aeb7-11e9-8b0e-f09c04faf3ab.gif)

chrisronline commented:
> I noticed that when I go to the Logstash Node Listing page the Logstash Overview tab disappears.

Legit bug. Fixed up now

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9386/)

chrisronline commented:
@ycombinator That issue should be addressed now and the new flows added to the PR description. Ready again for 👀

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9587/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9602/)

ycombinator commented:
The Logstash migration/setup functionality LGTM.

Testing Beats migration/setup, I noticed a couple of things in the steps:

> By default the module will collect beat monitoring metrics from http://localhost:5066. If the local beat instance has a different address, you must specify it via the hosts setting in the `modules.d/beat-xpack.yml` file.

Could we change the start of the second sentence to say, "If the beat instance being monitored has a...", since there's no strict requirement that the beat instance be local.

> In order for Metricbeat to collect metrics from the running beat, you need to [enable an HTTP endpoint]().

* Could we make "enable an HTTP endpoint" open in a new window?

* Could we change the wording to "enable the HTTP endpoint for the beat instance being monitored"?

chrisronline commented:
> Could we change the start of the second sentence to say, "If the beat instance being monitored has a...", since there's no strict requirement that the beat instance be local.

Done

> Could we make "enable an HTTP endpoint" open in a new window?

Done

> Could we change the wording to "enable the HTTP endpoint for the beat instance being monitored"?

Done

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9622/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9635/)

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/apm/instances/instances.js:
```diff
@@ -83,14 +83,37 @@ const columns = [
+          <p>
+            {i18n.translate('xpack.monitoring.apm.instances.metricbeatMigration.detectedInstanceDescription', {
+              defaultMessage: `Based on your indices, we think you might have an apm server. Click the 'Setup monitoring'
+              button below to start monitoring this apm server.`
```
Shouldn't all instances of `apm` be uppercase `APM`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/cluster/overview/apm_panel.js:
```diff
@@ -22,18 +21,55 @@ import {
+          content={i18n.translate('xpack.monitoring.cluster.overview.apmPanel.setupModeNodesTooltip', {
+            defaultMessage: `These numbers indicate how many detected monitored apm versus how many ` +
+            `detected total apm. If there are more detected apm than monitored apm, click the Nodes ` +
+            `link and you will be guided in how to setup monitoring for the missing node.`
```
Same as above maybe change `apm` tom `APM`

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/cluster/overview/apm_panel.js:
```diff
@@ -90,27 +128,32 @@ export function ApmPanel(props) {
+                      aria-label={i18n.translate(
+                        'xpack.monitoring.cluster.overview.apmPanel.instancesTotalLinkAriaLabel',
+                        {
+                          defaultMessage: 'Apm Instances: {apmsTotal}',
```
Maybe `APM` instead of `Apm`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/cluster/overview/beats_panel.js:
```diff
@@ -17,19 +17,56 @@ import {
+          content={i18n.translate('xpack.monitoring.cluster.overview.beatsPanel.setupModeNodesTooltip', {
+            defaultMessage: `These numbers indicate how many detected monitored beats versus how many ` +
+            `detected total beats. If there are more detected beats than monitored beats, click the Nodes ` +
+            `link and you will be guided in how to setup monitoring for the missing node.`
```
Maybe change `beats` to `Beats`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/cluster/overview/logstash_panel.js:
```diff
@@ -21,19 +21,58 @@ import {
+  const queueTypes = props.queue_types || {};
+
+  // Do not show if we are not in setup mode
+  if (!nodesCount && !setupMode.enabled) {
```
Should this maybe be `if (!nodesCount || !setupMode.enabled) {` so if only one of the conditions is false return null

chrisronline replied:
It needs to be both. If setup mode is enabled, we want to always show the panel, even if there are no nodes.

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/logstash/listing/listing.js:
```diff
@@ -123,17 +124,47 @@ export class Listing extends PureComponent {
+          >
+            <p>
+              {i18n.translate('xpack.monitoring.logstash.nodes.metribeatMigration.netNewUserDescription', {
+                defaultMessage: `Based on your indices, we think you might have a logstash node. Click the 'Setup monitoring'
```
Maybe capitalize `logstash` to `Logstash`

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/apm/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,267 @@
+          <EuiText>
+            <FormattedMessage
+              id="xpack.monitoring.metricbeatMigration.apmInstructions.metricbeatSecuritySetup"
+              defaultMessage="If security features are enabled, there may be more setup required.{link}"
```
Looks like you might need a space after the dot.

chrisronline replied:
There is a space in the link fragment actually that should get the job done, unless you're not seeing a space when running it?

igoristic replied:
Was just looking from the code point of view, but I understand now. Will test in a little bit 👍

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/apm/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,267 @@
+
+  const enableMetricbeatModuleStep = {
+    title: i18n.translate('xpack.monitoring.metricbeatMigration.apmInstructions.enableMetricbeatModuleTitle', {
+      defaultMessage: 'Enable and configure the beat x-pack module in Metricbeat'
```
Maybe capitalize `Beat`?

chrisronline replied:
I'm not sure in this case as the module is called `beat-xpack`. I can go either way I guess. Anyone else have thoughts here?

cachedout replied:
IMHO, `Beat` is still a proper name in the same way that we capitalize `Metricbeat`. Though, I will grant you that the naming scheme that was chosen makes this feel a little awkward. :)

chrisronline replied:
Changed!

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/apm/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,267 @@
+            size="s"
+            color="warning"
+            title={i18n.translate('xpack.monitoring.metricbeatMigration.apmInstructions.isInternalCollectorStatusTitle', {
+              defaultMessage: `We have not detected any monitoring data coming from Metricbeat for this Apm.
```
Maybe uppercase `APM`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/beats/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,307 @@
+          <EuiText>
+            <FormattedMessage
+              id="xpack.monitoring.metricbeatMigration.beatsInstructions.metricbeatSecuritySetup"
+              defaultMessage="If security features are enabled, there may be more setup required.{link}"
```
Might need space after the dot

chrisronline replied:
See my note from the other comment about needing a space after the dot

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/beats/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,307 @@
+                      >
+                        <FormattedMessage
+                          id="xpack.monitoring.metricbeatMigration.beatsInstructions.enableMetricbeatModuleHttpEnabledDirectionsLinkText"
+                          defaultMessage="enable an HTTP endpoint for the {beatType} instance being monitored"
```
Doesn't this need to start with a capital E? `Enable an...` and also end with a period?

chrisronline replied:
It's contained within a sentence already. Look at the `<FormattedMessage>` for `xpack.monitoring.metricbeatMigration.beatsInstructions.enableMetricbeatModuleHttpEnabledDirections` right above this

igoristic replied:
Ah, I see it now, thanks for clarifying!

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/beats/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,307 @@
+            size="s"
+            color="warning"
+            title={i18n.translate('xpack.monitoring.metricbeatMigration.beatsInstructions.isInternalCollectorStatusTitle', {
+              defaultMessage: `We have not detected any monitoring data coming from Metricbeat for this Beats.
```
Should `Beats` maybe be singular here `Beat`? or if plural maybe something like: `...from these Beats`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/logstash/disable_internal_collection_instructions.js:
```diff
@@ -0,0 +1,197 @@
+          <p>
+            <FormattedMessage
+              id="xpack.monitoring.metricbeatMigration.logstashInstructions.disableInternalCollection.note"
+              defaultMessage="You'll need to restart logstash after making this change"
```
Maybe capital `Logstash`?

igoristic commented on x-pack/legacy/plugins/monitoring/public/components/metricbeat_migration/instruction_steps/logstash/enable_metricbeat_instructions.js:
```diff
@@ -0,0 +1,267 @@
+          <EuiText>
+            <FormattedMessage
+              id="xpack.monitoring.metricbeatMigration.logstashInstructions.metricbeatSecuritySetup"
+              defaultMessage="If security features are enabled, there may be more setup required.{link}"
```
Space after dot?

chrisronline replied:
See my other comment about this

igoristic commented:
Everything looks good! Great job 💪

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10281/)

chrisronline commented:
Backport:

7.x: https://github.com/elastic/kibana/commit/1597b30132bf93da34fe79b3819979dd3384994d
